### PR TITLE
Create initial AdvancedSodium structure and implement crypto_scalarmult

### PR DESCRIPTION
**Pull request by timshadel**
As the first pass at the smallest portion of #5 I decided to implement `crypto_scalarmult`. Will you please look over my approach and see how well I managed to follow your idea to separate the advanced from the standard. Also, I'm really not sure how well I used the `sumo` term across both the `SodiumInit` implementations and the integration tests parameter.

I have _NOT_ implemented this in the SodiumJS area yet. I'd prefer to do that _in this PR_ but not until I get some preliminary feedback on the approach I've taken for the advanced API, file naming and locations, unit and integration testing approach, etc. At that point I'll implement JS and push to this PR for final review and merging.

### CODE REVIEW

File: packages/sodium/lib/src/api/advanced/advanced_crypto.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import '../crypto.dart';
+import 'advanced_scalar_mult.dart';
+
+abstract class AdvancedCrypto implements Crypto {
+  const AdvancedCrypto._(); // coverage:ignore-line
+
+  // @override
+  // AdvancedSecretStream get secretStream;
+
+  AdvancedScalarMult get scalarMult;
```
Comment (Skycoder42) [inline packages/sodium/lib/src/api/advanced/advanced_crypto.dart:10]: Documentation is missing here

File: packages/sodium/lib/src/ffi/sodium_ffi_init.dart [hunk 2]
```diff
@@ -45,4 +47,19 @@ abstract class SodiumInit {
...
   }
+
+  // coverage:ignore-start
+  /// Creates an [AdvancedSodium] instance for the loaded libsodium with
+  /// both standard and advanced features available.
+  ///
+  /// The [libsodium] parameter must be a loaded
+  /// `[lib]sodium.[so|dll|dylib|a|lib|js]`- depending on your platform. Please
+  /// refer to the README for more details on loading the library.
+  static Future<AdvancedSodium> initSumo(DynamicLibrary libsodium) {
```
Comment (Skycoder42) [inline packages/sodium/lib/src/ffi/sodium_ffi_init.dart:58]: Why is this here? Doesn't  advanced_sodium_ffi_init.dart already do this?

**Review by Skycoder42 — CHANGES_REQUESTED** [review]
Firstoff: Thanks for the PR. Generally, it looks very good. And sorry for the late reply, but those are quite a few changes, so I had to take my time to look at it. I am fine with this beening just the FFI part, and yes, I would also prefer the JS part to be added to the PR. I have turned the PR into a draft for now, so signify it not beeing ready to be merged yet. Now to the actual review:

Regarding the scalarmult implementation, it looks just as I would have expected it. There are a few small comments, but nothing big. I also like the fact that you tested for "no more interactions" after the tests. Integration tests look fine too, though I have not run them myself yet.

Next, the general advanced architecture. It follows the Idea I had in #5, but the fact that we have to redeclare the `late final Crypto _crypto;` like that is kinda ugly... However, I was not able to find a better solution yet, as this keeps the public interface clean and thanks to late initialization ensures the the overridden member never gets initialized, so I guess it is fine.  

I also like the creation of `sodium.sumo.dart`, as that keeps the two parts cleanly separated with an explicit opt-in to the sumo variants. However, I am not sure about the changes you did to the initialization methods. Here are my somewhat unstructured thoughts:
- `advanced_sodium_init_fallback.dart` should extends the normal advanced fallback
- Same for `advanced_sodium_ffi_init.dart` - it should extends the normal ffi implementation
- the initSumo should not be present in `sodium_ffi_init.dart`
- the `advanced_..._init....dart` should be named sumo instead.

Regarding the whole sumo/advanced naming: The libsodium documentation calls these APIs "advanced", they are listed under that tab and are referred as such. However, the sodium.js repository (by the same author) differentiates between "standard" and "sumo" - with "standard" beeing a JS lib with only normal APIs, and "sumo" beeing the full thing. This is why I used both in my draft. The reasoning was:
- The APIs themselves are referred to as "advanced" - so the classe should be named like that
- The library as a whole is reffered to as "normal" or "sumo" - so "entry points", like the library import or the initialization part should be named sumo.

I hope this clears it up for you a little - and makes sense to you. If you think this is stupid and we should just settle on one of the two, please feel free to tell me so.

Regarding the integration tests, I think sumo tests should always run all tests, not just the new ones, just to ensure everything always works. Also, stuff should be renamed, as detailed in the actual comments. Have a look at the JS integration test setup, it might help you to get the structure right. But that is not a requirement, I can clean it up afterwards.

Now about the pipeline: It failed, simply, because the dart SDK has been updated and now the linter finds issue that have not been there before. If you want to, you can add the following  as an extra commit to your PR. That should satisfy the linter. I can also do the commit myself, but then you would have to rebase you PR - whatever you like better. 

```.yaml
// Modify packages/sodium/analysis_options.yaml, add the following to the bottom:

    # temporarily disabled
    require_trailing_commas: false
    avoid_classes_with_only_static_members: false
```

Once the CI works again, we would also have to add a configuration for running the sumo tests in the CI. However, unless you have experience with GitHub Actions and want to do it yourself, I would simply do that myself as "final step" of the PR, so don't worry about that.

**Comment by timshadel** [comment]
I like your direction on all the comments, and I'll dive into them over the weekend. I'll plan to rebase, if you don't mind making that commit. I strongly prefer allowing `main` to progress and rebasing PRs to ensure compatibility and clean history before final merges.

File: packages/sodium/lib/src/ffi/api/advanced/advanced_sodium_ffi.dart [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import '../../../api/advanced/advanced_sodium.dart';
+import '../../bindings/libsodium.ffi.dart';
+import '../sodium_ffi.dart';
+import 'advanced_crypto_ffi.dart';
+
+class AdvancedSodiumFFI extends SodiumFFI implements AdvancedSodium {
+  AdvancedSodiumFFI(LibSodiumFFI sodium) : super(sodium);
+
+  @override
+  AdvancedCrypto get crypto => _crypto;
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/lib/src/ffi/api/advanced/advanced_sodium_ffi.dart:11]: oh, I meant to name this one `_advancedCrypto`. As we only need to change out the getter response there's no need to have the private vars collide one over the other. Does that resolve your question here? The major change is in `sodium_ffi.dart` where we move `crypto` from a field that others can see to a getter that subclasses can override.

**Comment by Skycoder42** [comment]
All updates to main are pushed. The sodium builds already work, the flutter library failed on windows, but should work now after my last push. You can rebase your branch and continue with it. Sorry it took so long. Have a nice day!

File: packages/sodium/test/integration/advanced_test_case.dart [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import 'package:meta/meta.dart';
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/integration/advanced_test_case.dart:1]: This was my part of my first attempt, and it just didn't get deleted with the original.

File: packages/sodium/test/integration/vm_standard_test.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import 'dart:async';
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/integration/vm_standard_test.dart:1]: I went the other way and moved the js one to "js_test_standard.dart"; the parallelism between "standard" and "sumo" seems to be informative and descriptive. "vm_test.dart" seems like it's generic enough it should be calling both standard and sumo. After you see it, if you really want it to be "vm_test.dart" I'll change it all back.

File: packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+          () => mockSodium.sodium_allocarray(5, 1),
+          () => mockSodium.sodium_mprotect_noaccess(
+                any(that: isNot(nullptr)),
+              ),
+          () => mockSodium.sodium_allocarray(5, 1),
+          () => mockSodium.sodium_mprotect_readonly(
+                any(that: isNot(nullptr)),
+              ),
+          () => mockSodium.sodium_mprotect_readonly(
+                any(that: isNot(nullptr)),
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart:198]: One of these changes worked (above), but the rest gave me output like

```
00:01 +12 -1: loading test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart                                                                                                                     
HasRawDataMatcher failed with: Expected: <Instance of 'Pointer<Uint8>'>
  Actual: Pointer<Void>:<Pointer<Void>: address=0x7f9c2dc04240>
   Which: is not an instance of 'Pointer<Uint8>'
```

I'm not quite familiar enough with the setup to know what that implies. Can you give me another hint?

File: packages/sodium/lib/src/ffi/sodium_ffi_init.dart [hunk 7]
```diff
@@ -1,7 +1,9 @@
 import 'dart:ffi';
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/lib/src/ffi/sodium_ffi_init.dart:1]: Maybe I missed what this comment was referring to...

**Comment by timshadel** [comment]
Ok, I've made changes and rebased on master. I've included changes to the CI items to run the sumo tests (but please check my work). I've resolved most of the comments, but need a bit more detail on a couple. Let me know what you think. When it looks good I'll start on the JS side.

**Comment by Skycoder42** [comment]
Very nice, besides of the remaining comments, I think this looks good, you can start the JS side. The only thing missing is the adjustment of the CI. Instead of running both sumo and non sumo variants, you should change it as stated in the comment and adjust the github actions pipeline, so that one separate job is run for each platform and each variant. However, is can become quite complicated, so here is my idea (after merging the PR, I would the try to "clean" the CI, but this is a trial and error process). I created a diff you can apply to the `.github/workflows/sodium_ci.yaml`

```.diff
diff --git a/.github/workflows/sodium_ci.yaml b/.github/workflows/sodium_ci.yaml
index 3005c8a..e1f975d 100644
--- a/.github/workflows/sodium_ci.yaml
+++ b/.github/workflows/sodium_ci.yaml
@@ -38,12 +38,15 @@ jobs:
           - platform: linux
             os: ubuntu-latest
             system: vm
+            integrationTestArgs: --sumo
           - platform: windows
             os: windows-latest
             system: vm
+            integrationTestArgs: --sumo
           - platform: macos
             os: macos-latest
             system: vm
+            integrationTestArgs: --sumo
           - platform: web
             os: ubuntu-latest
             system: js
```

This would ensure, the VM platforms always run the Sumo tests, while JS is kept as is

File: packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart:198]: @Skycoder42 can you give me another idea; that one gave the above error

File: packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+          () => mockSodium.sodium_mprotect_readonly(
+                any(that: hasRawData(secretKey)),
+              ),
+          () => mockSodium.crypto_scalarmult(
+                any(that: isNot(nullptr)),
+                any(that: hasRawData<Uint8>(secretKey)),
+                any(that: isNot(nullptr)),
+              ),
+          () => mockSodium.sodium_free(
+                any(that: isNot(nullptr)),
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart:210]: @Skycoder42 see above error

File: packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+              ),
+          () => mockSodium.sodium_mprotect_noaccess(
+                any(that: isNot(nullptr)),
+              ),
+          () => mockSodium.sodium_free(
+                any(that: isNot(nullptr)),
```
[the remark this answers is not in the document]
Comment (timshadel) [reply packages/sodium/test/unit/ffi/api/advanced/advanced_scalarmult_ffi_test.dart:216]: @Skycoder42 see above error

**Comment by timshadel** [comment]
@Skycoder42 I've run into 2 issues which have me stuck. First, I can't run web integration tests on my machine because they rely on code in a `bindings/` directory of the integration tests which has a specific `.gitignore` keeping it out of git (and I find no mention of it in the docs). Second, I can't see the output of the build actions right now, so I can't see the error occurring with the web sumo build. All action details (for _both_ success and failed steps) just shows a blank error. Dev tools show that some URL with an S3 presigned URL is failing. It seems like something on Github is broken at the moment.

Screen Shot 2021-10-17 at 6 06 33 PM

Would you mind looking at the issue when you have a minute? I think that the implementation is pretty well done, but until I can see all the tests pass I'm not quite convinced. I could use your review of the test adaptations I made for JS, in addition to checking my implementation code.

File: packages/sodium/lib/src/js/sodium_js_sumo_init.dart [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import '../api/advanced/advanced_sodium.dart';
+import 'api/advanced/advanced_sodium_js.dart';
+import 'bindings/sodium.js.dart';
+
+/// Static class to obtain a [Sodium] instance.
+///
+/// **Important:** This API is is different depending on whether it is used from
+/// the VM or in transpiled JavaScript Code. See the specific implementations
+/// for more details.
+abstract class SodiumInit {
```
Comment (Skycoder42) [inline packages/sodium/lib/src/js/sodium_js_sumo_init.dart:12]: Update documentation for the whole file. Also, the simple "initSumo" is missing.

Do you think it would be useful to add a simply test, if the given JS object is actually a sumo instance? for example, by checking if a certain sumo API is available?

File: packages/sodium/lib/src/js/sodium_js_sumo_init.dart [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  const SodiumInit._(); // coverage:ignore-line
+
+  /// Creates a [Sodium] instance for the loaded libsodium as [LibSodiumJS].
+  ///
+  /// The [sodiumJsObject] parameter must be a loaded `sodium.js` -
+  /// depending on your platform. Please refer to the README for more details
+  /// on loading the library.
+  static Future<AdvancedSodium> initFromSodiumJS(LibSodiumJS libsodium) =>
```
Comment (Skycoder42) [inline packages/sodium/lib/src/js/sodium_js_sumo_init.dart:20]: rename to `initFromSodiumJSSumo`?

File: packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        ).thenReturn(Uint8List(0));
+
+        final secretKey = List.generate(5, (index) => index);
+
+        sut.base(secretKey: SecureKeyFake(secretKey));
+
+        verifyInOrder([
+          () => mockSodium.crypto_scalarmult_SCALARBYTES,
+          () => mockSodium.crypto_scalarmult_base(
+                any(),
```
Comment (Skycoder42) [inline packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart:76]: Pass `secretKey` instead of `any()` here to verify that the correct key was passed to base

File: packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        verifyNoMoreInteractions(mockSodium);
+      });
+
+      test('returns public key', () {
+        final publicKey = List.generate(5, (index) => index);
+        when(
+          () => mockSodium.crypto_scalarmult_base(
+            any(),
+          ),
+        ).thenAnswer((_) => Uint8List.fromList(publicKey));
```
Comment (Skycoder42) [inline packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart:88]: `thenAnswer` is okay here, but since this is a synchronous call, you can actually use `.thenReturn(Uint8List.fromList(publicKey))`

File: packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        sut.call(
+          secretKey: SecureKeyFake(secretKey),
+          otherPublicKey: Uint8List.fromList(otherPublicKey),
+        );
+
+        verifyInOrder([
+          () => mockSodium.crypto_scalarmult_SCALARBYTES,
+          () => mockSodium.crypto_scalarmult_BYTES,
+          () => mockSodium.crypto_scalarmult(
+                any(),
```
Comment (Skycoder42) [inline packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart:158]: Same as above. Use `secretKey` and `otherPublicKey`

File: packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+      });
+
+      test('returns shared secret', () {
+        final sharedSecret = List.generate(5, (index) => index);
+        when(
+          () => mockSodium.crypto_scalarmult(
+            any(),
+            any(),
+          ),
+        ).thenAnswer((_) => Uint8List.fromList(sharedSecret));
```
Comment (Skycoder42) [inline packages/sodium/test/unit/js/api/advanced/advanced_scalarmult_js_test.dart:172]: Same, you can use `thenReturn`

**Comment by timshadel** [comment]
Wow, I lost track of time! I may need your help with the PR, but I am still having fun with it. :) I'll let you know if it gets out of hand or life takes over again.

**Comment by Skycoder42** [comment]
Hi^^ Since you seem to have lost the time to keep working on this, I would like to try and "finish" the PR step by step. For now, my first goal would be to recreate the basics for the "advanced" APIs. Is that fine with you? Since I am going to base my changes on your work so far, I will of course mention you in these commits/releases

**Comment by Skycoder42** [comment]
Okay, so I have implemented the sumo-APIs, as well as the scalarmult APIs with version 2.0.0

With that, this PR becomes invalid and thus I will close it.

**Comment by timshadel** [comment]
Thank you!!!
